### ISSUE TITLE
Version string in released binaries is missing the "v"

### ISSUE THREAD

**Issue by johananl** [kinvolk/lokomotive#1150]
The binaries we release in official releases print a different version string than binaries built from the same Git tag using `make`. Building a binary from a certain Git revision should have a deterministic version string baked into it. It shouldn't matter exactly which command was used to build the binary.

Steps to reproduce:

```
wget https://github.com/kinvolk/lokomotive/releases/download/v0.5.0/lokoctl_0.5.0_linux_amd64.tar.gz
tar zxvf lokoctl_0.5.0_linux_amd64.tar.gz
./lokoctl_0.5.0_linux_amd64/lokoctl version
```

Output: `0.5.0`

```
git clone https://github.com/kinvolk/lokomotive.git
cd lokomotive
git checkout v0.5.0
make
./lokoctl version
```

Output: `v0.5.0`

@iaguis @surajssd - your fingerprints are around this part of the codebase so tagging you for attention :slightly_smiling_face:

**Comment by invidian** [comment]
This is because we use 
```
-ldflags "-X github.com/kinvolk/lokomotive/pkg/version.Version=`git describe --tags --always`
```
 in the Makefile.

I think if we tag as `v0.5.0`, `lokoctl version` should also print `v0.5.0`.

**Comment by johananl** [comment]
Don't we use the same `-ldflags` trick when crafting releases? I see no reason for using one process for dev builds and another for prod builds. I suspect this is somehow related to the fact that we use `goreleaser` when releasing.

**Comment by invidian** [comment]
You're right. We do use `-ldflags` when building release builds. The difference probably comes from the fact that `{{.Version}}` macro from `goreleaser` does not match output from `git describe --tags --always`.

`{{.Version}}` [documentation](https://goreleaser.com/customization/templates/) explicitly says that `v` will be stripped. Perhaps we should use `{{.Tag}}` instead then.

**Comment by invidian** [comment]
Hm, after re-visiting it, I have concerns about including `v` in printed version. `v` in tag indicates that it's a **v**ersion of the release. Looking at `git` for example, they use `v2.29.2` and `2.29.2` interchangeably.

```console
$ git --version
git version 2.29.2
```

I may actually consider admitting, that current behavior is actually correct. Release build prints `0.5.0`, if you checkout the `v0.5.0` tag and do a development build, you get `v0.5.0` printed. But it is not the release build as would be done with `goreleaser`.

I wonder if `goreleaser` produces reproducible builds...

**Comment by johananl** [comment]
`v` usually means "version", so if you say `git version v2.29.2` you are saying "version" twice. `git v2.29.2` isn't redundant IMO.

Personally I like the `v`, but officially the `v` isn't part of the semver string IIRC. Anyway, I'm fine either way as long as we are consistent, and as said above I think we should use the same process for dev builds and release builds to avoid this kind of mismatches.

**Comment by iaguis** [comment]
Let's create a Makefile tag that uses goreleaser to generate a release but not upload it to GitHub.

**Comment by iaguis** [comment]
This will happen in the v0.7.0 PR.

**Comment by invidian** [comment]
Suggestion for person implementing it: maybe not make a target name `release`. Maybe `build-release` running `goreleaser release --skip-publish --skip-sign`?

**Comment by johananl** [comment]
I'd go with `make release` for crafting a release locally and then `make publish` for publishing it.

**Comment by invidian** [comment]
We already have `build-and-publish-release`, which does both. `goreleaser` does not have separate `publish` step.

**Comment by johananl** [comment]
> Let's create a Makefile tag that uses goreleaser to generate a release but not upload it to GitHub.

@iaguis I'm suddenly not sure why it would be useful to have one Make taget which generates a release and another which uploads it. What would be the use case for only generating a release and not uploading it?

My original concern with this issue was inconsistency between a released binary and the one resulting from running `make` against the relevant Git tag. AFAICT we should have *one* target which generates a binary, and that target should be reused both when just building a binary and when publishing a release. I feel like goreleaser might be getting in our way here.

Maybe we should continue this discussion in a separate issue and merge just the "v" fix in order not to block v0.7.0?

**Comment by iaguis** [comment]
The point with having a make target that generates a release was to provide users a way to build the lokoctl binary with the same behavior as official releases. That is, that `lokoctl version` returns `0.x.x` without the `v`.

Having thought about it and talked with several ppl in the team I now think we should change the way we display the version in `lokoctl version` to `v0.x.x`. The reasoning, like you point out in https://github.com/kinvolk/lokomotive/issues/1150#issuecomment-724849672 is that the actual version is `0.x.x` and the string `v` means "version".

I think we should add the `v` in the version string like https://github.com/kinvolk/lokomotive/pull/1417 which should fix the user-facing issue and then we can worry about having two different ways of building lokoctl (with make and with goreleaser) after releasing v0.7.0.

**Comment by meyskens** [comment]
One case to be made for the new build target is to have builds reproducible to the level op the exact cross-compile and CGO settings of goreleaser making a Make command for having all release artifacts without publishing.

however on the other side there may be no demand for that making it a unused option.

**Comment by iaguis** [comment]
Yes, I think that's a separate issue to make local builds and releases have the same compile settings. This issue can be closed with #1147.

**Comment by invidian** [comment]
>Yes, I think that's a separate issue to make local builds and releases have the same compile settings. This issue can be closed with #1147.

Goreleaser already use `CGO_ENABLED=0` and `-extldflags '-static'`, so they use the same setting? or you mean to deduplicate them?

**Comment by iaguis** [comment]
> Goreleaser already use `CGO_ENABLED=0` and `-extldflags '-static'`, so they use the same setting? or you mean to deduplicate them?

Yes, to avoid mistakes when changing them.

**Comment by johananl** [comment]
Let's move the discussion around the separate Make targets to https://github.com/kinvolk/lokomotive/issues/1418. This issue will be closed once #1417 is merged.
